Changelog 144 (#112)
intellij-rust/intellij-rust.github.io: _posts/2021-03-29-changelog-144.markdown

---
layout: post
title: "IntelliJ Rust Changelog #144"
date: 2021-03-29 17:00:00 +0300
---

## New Features

<!-- https://github.com/intellij-rust/intellij-rust/pull/6956 -->
* Increase minimal supported rustc version to 1.41.0

<!-- https://github.com/intellij-rust/intellij-rust/pull/6907 -->

> * Provide debugging with LLDB in [Rider for Unreal Engine](https://www.jetbrains.com/lp/rider-unreal/) 2021.1

Reviewer: I suppose these changes are quick minor so we should move them below

> * Provide debugging with LLDB in [Rider for Unreal Engine](https://www.jetbrains.com/lp/rider-unreal/) 2021.1

Reviewer: Let's also add note that `LLDB` on Windows means only MSVC toolchain

<!-- https://github.com/intellij-rust/intellij-rust/pull/6799 -->
* Complete dependencies by subwords in `Cargo.toml` when using experimental local crates index. This feature is disabled by default

>   for now. To use it, enable `org.rust.crates.local.index` experimental feature

Reviewer: Let's add a note on how to enable experimental feature. It's quite frequent user question

  {% include gif-img.html path="/assets/posts/changelog-144/6799" w="600px" %}

<!-- https://github.com/intellij-rust/intellij-rust/pull/6893 -->
* Import unresolved types from a function signature after performing `Extract Function` refactoring (`Refactor | Extract Method` or
  <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>M</kbd>)

<!-- https://github.com/intellij-rust/intellij-rust/pull/6947 -->

> * Support import aliases in `Move` refactoring (`Refactor | Move` or <kbd>F6</kbd>)

Reviewer: GIF?

<!-- https://github.com/intellij-rust/intellij-rust/pull/6966 -->
* Automatically select the default project Rust toolchain from `Preferences | Languages & Frameworks | Rust` when creating a new project

<!-- https://github.com/intellij-rust/intellij-rust/pull/7001 -->

> * (maybe group together with other const generic things?) Support [const generics](https://rust-lang.github.io/rfcs/2000-const-generics.html) in `Wrong generic arguments number` ([E0107](https://doc.rust-lang.org/error-index.html#E0107)) inspection

Author: @Undin WDYT about grouping together all the const generic items, in either Features or Fixes block?

Reviewer: I don't mind

Author: I moved all const generic items into `New Features` block

<!-- https://github.com/intellij-rust/intellij-rust/pull/6066 -->

> *  Use [`todo!()`](https://doc.rust-lang.org/std/macro.todo.html) instead of [unimplemented!()](https://doc.rust-lang.org/std/macro.unimplemented.html) in function bodies generated by `Implement members` action (by [@KarelPeeters])

Reviewer: ```suggestion
*  Use [`todo!()`](https://doc.rust-lang.org/std/macro.todo.html) instead of [`unimplemented!()`](https://doc.rust-lang.org/std/macro.unimplemented.html) in function bodies generated by `Implement members` action (by [@KarelPeeters])
```

Reviewer: Not only by `Implement members`

<!-- https://github.com/intellij-rust/intellij-rust/pull/6934 -->
* Auto-indent line when typing <kbd>Enter</kbd> in macro bodies

<!-- https://github.com/intellij-rust/intellij-rust/pull/6912 -->
* Enable by default `Simplify boolean expression` inspection

## Fixes

<!-- https://github.com/intellij-rust/intellij-rust/pull/6959 -->

> * (maybe feature?) Support `Move Element Right/Left` actions for const generic parameters

Reviewer: Yep, let's move it into feature section

---
layout: post
title: "IntelliJ Rust Changelog #144"
date: 2021-03-29 17:00:00 +0300
---

## New Features

<!-- https://github.com/intellij-rust/intellij-rust/pull/6799 -->

> * Complete dependencies by subwords in `Cargo.toml` when using experimental local crates index. This feature is disabled by default

Reviewer: Let's reorder items a little bit.
I suggest the following order: import aliases in `Move` refactoring, const generics, `Complete dependencies` and other in current order

Author: I like this idea. Actually, I put this item on the top because of an attractive gif. But now we have another gif about `Move` refactoring 👌

* Import unresolved types from a function signature after performing `Extract Function` refactoring (`Refactor | Extract Method` or
  <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>M</kbd>)
  
* [Const generics](https://rust-lang.github.io/rfcs/2000-const-generics.html) support improvements:
  <!-- https://github.com/intellij-rust/intellij-rust/pull/7001 -->
  * Support const generics in `Wrong generic arguments number` ([E0107](https://doc.rust-lang.org/error-index.html#E0107)) inspection

  <!-- https://github.com/intellij-rust/intellij-rust/pull/6959 -->
  * Support `Move Element Right/Left` actions for const generic parameters

  <!-- https://github.com/intellij-rust/intellij-rust/pull/6958 -->
  * Properly render const generics in signature of a top-level item
  
  <!-- https://github.com/intellij-rust/intellij-rust/pull/6948 -->
  * Fix type rendering with const generic parameters

<!-- https://github.com/intellij-rust/intellij-rust/pull/6966 -->
* Automatically select the default project Rust toolchain from `Preferences | Languages & Frameworks | Rust` when creating a new project

<!-- https://github.com/intellij-rust/intellij-rust/pull/6066 -->

> *  Use [`todo!()`](https://doc.rust-lang.org/std/macro.todo.html) instead of [`unimplemented!()`](https://doc.rust-lang.org/std/macro.unimplemented.html) in function bodies generated by `Implement members`, `Create function`, and `Replace .unwrap() with match` actions (by [@KarelPeeters])

Reviewer: ```suggestion
*  Use [`todo!()`](https://doc.rust-lang.org/std/macro.todo.html) instead of [`unimplemented!()`](https://doc.rust-lang.org/std/macro.unimplemented.html) in function bodies by the plugin's code generation features like  `Implement members` and `Create function` actions (by [@KarelPeeters])
```

WDYT?

Author: Looks good! Maybe mention https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-generation.html doc page as well?

Reviewer: Good idea!

Author: BTW I'm not sure that "plugin" is necessary here. For me, "by [code generation](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-generation.html) features" is clear enough

Reviewer: It's up to you here

<!-- https://github.com/intellij-rust/intellij-rust/pull/6934 -->
* Auto-indent line when typing <kbd>Enter</kbd> in macro bodies

<!-- https://github.com/intellij-rust/intellij-rust/pull/6907 -->

> * Provide debugging with LLDB in [Rider for Unreal Engine](https://www.jetbrains.com/lp/rider-unreal/) 2021.1. Note, only MSVC toolchain is supported

Reviewer: ```suggestion
* Provide debugging with LLDB in [Rider for Unreal Engine](https://www.jetbrains.com/lp/rider-unreal/) 2021.1. Note, only [MSVC toolchain](https://doc.rust-lang.org/edition-guide/rust-2018/platform-and-target-support/msvc-toolchain-support.html) is supported
```
